pyiron_base/job/template.py

```diff
@@ -23,29 +24,62 @@
 class TemplateJob(GenericJob):
     def __init__(self, project, job_name):
         super().__init__(project, job_name)
-        self.input = GenericParameters(table_name="input")
+        self._input = InputList(table_name="input")
+
+    @property
+    def input(self):
+        return self._input
+
+    @input.setter
+    def input(self, input):
+        self._input = InputList(input, table_name="input")
 
     def to_hdf(self, hdf=None, group_name=None):
         super().to_hdf(
             hdf=hdf,
             group_name=group_name
         )
-        with self.project_hdf5.open("input") as h5in:
-            self.input.to_hdf(h5in)
+        self._input.to_hdf(hdf=self._hdf5, group_name=None)
 
     def from_hdf(self, hdf=None, group_name=None):
         super().from_hdf(
             hdf=hdf,
             group_name=group_name
         )
-        with self.project_hdf5.open("input") as h5in:
-            self.input.from_hdf(h5in)
+        self._input.from_hdf(hdf=self._hdf5, group_name=None)
 
 
 class PythonTemplateJob(TemplateJob):
     def __init__(self, project, job_name):
         super().__init__(project, job_name)
+        self._output = InputList(table_name="output")
         self._python_only_job = True
 
+    @property
+    def output(self):
+        return self._output
+
+    @output.setter
+    def output(self, output):
+        self._output = InputList(output, table_name="output")
+
+    def from_hdf(self, hdf=None, group_name=None):
+        super().from_hdf(
+            hdf=hdf,
+            group_name=group_name
+        )
+        self._output.from_hdf(hdf=self._hdf5, group_name=None)
+
+    def to_hdf(self, hdf=None, group_name=None):
+        super().to_hdf(
+            hdf=hdf,
+            group_name=group_name
+        )
+        self._output.to_hdf(hdf=self._hdf5, group_name=None)
+
+    def save_output(self):
```

Comment:
I really share Marvin's concern about this.

What is the use case? Who is calling this when?

Overall I am a bit concerned about the job flow. Like, is `run_static` ever called by the end user, or from another class as opposed to internally by the job itself? It should, perhaps, be private? 

And if there's stuff that should *always* be happening after "user defined" `run_static` definitions, is it not also the case that it should always be happening, it would be nice to tuck this away elsewhere in the flow so that both users and developers only have to define a `run_static` without worrying about book keeping.

These issues might be beyond the scope of this PR, but it really looks to me like we start pushing up against these issues now.

Reply:
AFAIK `run_static` is never called by the user. It is always handled by the `run()` call, which calls all the `_run_if...` functions depending on the job status. This is the reason we have to specify  `self.status.finished = True` in the `run_static` method...
Actually we could also set `self.status.collect = True` in order to trigger  `_run_if_collect`  where we could handle the `_output.to_hdf`. However, we would be left with this `self.status.collect = True` poping up 'randomly' for a new developer.

Reply:
Yeah, I'm wondering if we should have something like

```python

def run_static(self):
    raise NotImplementedError

def _run_static(self):
    self.run_static()
    self.status.collect = True

def collect_output(self):
    pass

def _collect_output(self):
    self.collect_output()
    self.status.finished = True
```

As I write this, I immediately see why @jan-janssen was talking elsewhere about introducing a decorator... like

```python
@status_method('collect')
def calc_static:
    raise NotImplementedError
```

I share his concern about decorator complexity. Namely, I worry about forcing new devs to remember to add the decorator when they're doing the definition. But maybe it's just a fair cognitive load to add for the beauty we get in return.

Finally, if you're right that `run_static` and `collect_output` are never called from outside their class, then we should really make them private. Right now they crowd the tab completion list on job objects. This is a general problem we need to work on.

Reply:
I just realized, we have to call `self.run()` in the `_run_static` method to trigger the `collect_output()` method. I think it would be nicer to leave the `collect_output` method for file I/O where needed and to handle everything in `run_static()` as it is done right now.

Reply:
It is the other way around, the `run()` method calls `collect_output` if the status is set to collect: 
https://github.com/pyiron/pyiron_base/blob/master/pyiron_base/job/generic.py#L697
More precisely it calls `_run_if_collect()`: 
https://github.com/pyiron/pyiron_base/blob/master/pyiron_base/job/generic.py#L1483
Which then calls the collect output.

Reply:
> I just realized, we have to call `self.run()` in the `_run_static` method to trigger the `collect_output()` method.

@niklassiemer yes, I missed re-calling `run`. This is actually kind of my point; my understanding of the current setup is that the simulation should always flow through a series of steps controlled by its status, but as it stands right now developers still need to actively manage this flow in each subclass of jobs. The code is functional as is, so I don't want to come down too hard on it, I just feel like there is an opportunity here to really make strong gains in improving maintainability and extensibility by restructuring this flow process.

>  I think it would be nicer to leave the `collect_output` method for file I/O where needed and to handle everything in `run_static()` as it is done right now.

I still disagree here; if we have a `collect` phase and a `collect_output` method, I find it very strange to do data collection and post-processing in the `run` phase. It also shows up here a little bit that we have a whole extra method `save_output`. I worry that we are getting many similarly named methods, and still sticking functionality in methods with names that don't describe that functionality. Again, there's nothing wrong with how it operates, I just feel like it makes maintenance and further extension a headache.

Reply:
> It is the other way around, the `run()` method calls `collect_output` if the status is set to collect:
> https://github.com/pyiron/pyiron_base/blob/master/pyiron_base/job/generic.py#L697
> More precisely it calls `_run_if_collect()`:
> https://github.com/pyiron/pyiron_base/blob/master/pyiron_base/job/generic.py#L1483
> Which then calls the collect output.

@jan-janssen but is it not correct that even if we set `self.status.collect = True` in `self.run_static`, that without *also* calling `self.run()` inside `run_static`, we never proceed to the collection?

Reply:
@liamhuber Sorry for the confusing yes you should call run, just like it is also done in the main `run_static` method for external executables https://github.com/pyiron/pyiron_base/blob/master/pyiron_base/job/generic.py#L787

Reply:
@jan-janssen right, so this is a little bit my point; basically every class that inherits from `GenericJob` and is overriding `run_static` then needs to be adding these lines to set the status and re-call run. I think it's fine that this is the way we *have done it so far*, but now that we have people who are explicitly hired as software developers I'd like to take our existing "reasonable solutions" and level them up to "excellent solutions".

So in this case, the `run` method should have some sort of `while self.status not in [${list of terminating statuses}]: ${keep running}` and the downstream exposed `run_static` and `collect_output` methods should be wrapped up so that we don't need to keep re-implementing the status updates and run continuation. 

I'm totally open to new ideas for the details of this implementation, and also open to the timing of when we get around to doing it, but I hope we can come to (or are already at) a consensus that some sort of better packaging of the execution flow is important.

Reply:
@liamhuber The pyiron jobs were originally designed to call external codes and `run_static()` is used in the job wrapper and the GernericMasters so I am a bit hesitant to change everything. As far as I understand only the python based jobs need these changes so if we implement a function like `run_python_job(self)` which is called inside the `run_static()` function we should be able to maintain both functionality and smoothen the transition.

Reply:
A current use case is the current hessian job https://github.com/pyiron/pyiron/blob/master/pyiron/thermodynamics/hessian.py#L44 or the jobs derived from ASE https://github.com/pyiron/pyiron/blob/master/pyiron/gpaw/pyiron_ase.py#L125 these implement the `run_static()` function directly as the input and output handling is primarily using the interactive interface.

Reply:
> @liamhuber The pyiron jobs were originally designed to call external codes and `run_static()` is used in the job wrapper and the GernericMasters so I am a bit hesitant to change everything. As far as I understand only the python based jobs need these changes so if we implement a function like `run_python_job(self)` which is called inside the `run_static()` function we should be able to maintain both functionality and smoothen the transition.

@jan-janssen I am still struggling here. I think this is just proving my point -- historically we used pyiron jobs to wrap external executables, but now many of the newly written classes that are children of `GenericJob` *don't do that*. That includes the two examples you gave, the Generic Masters you mentioned, as well as [several](https://github.com/pyiron/pyiron_contrib/blob/569e059a28c1e68c58a68c65c4f8d1486c65bcc7/pyiron_contrib/atomistic/atomistics/master/sqselastic.py#L321) [examples](https://github.com/pyiron/pyiron_contrib/blob/71ba4eea9db6acebffd5fb9876b54e4268cd7aba/pyiron_contrib/protocol/generic.py#L639) [in](https://github.com/pyiron/pyiron_contrib/blob/71ba4eea9db6acebffd5fb9876b54e4268cd7aba/pyiron_contrib/image/job.py#L149) pyiron_contrib. So `GenericJob.run_static` is getting overwritten all the time, and in many of those cases the logic of `GenericJob.run_static` -- e.g. status setting and recursively calling `run()` -- needs to be re-implemented.

So `GenericJob` probably isn't the perfect parent for all these classes! Or at least this implementation of it isn't. What I'm asking for is to re-write the execution flow of a base-level `Job` class which other classes can inherit from and not need to re-do this logic all the time. This seems to be a bit the direction you're going with `run_python_job`, but also your reply sounded like you were disagreeing with me.

We don't have to do this all at once, we can make these changes incrementally. E.g. with taking a step to add `run_python_job` as you suggest, or to make `run_static` and `collect_output` private since they're never(?) called from outside themselves, etc. 

I get that overall this might not be a trivial change, but my point is that we now have a whole team of people who are explicitly software developers. I think we have the time to start gunning for harder problems so that our code can go from being 'historical' to 'optimal'. I'm not trying to nail down details, I am just hoping to get an agreement in principle.